aeppl/truncation.py

```diff
@@ -122,3 +138,237 @@ def censor_logprob(op, values, base_rv, lower_bound, upper_bound, **kwargs):
+    upper = at.as_tensor_variable(upper) if upper is not None else at.constant(np.inf)
+
+    if srng is None:
+        srng = RandomStream()
+
+    # Try to use specialized Op
+    try:
+        truncated_rv, updates = _truncated(
+            rv.owner.op, lower, upper, srng, *rv.owner.inputs[1:]
+        )
+        return truncated_rv, updates
+    except NotImplementedError:
+        pass
+
+    # Variables with `_` suffix identify dummy inputs for the OpFromGraph
+    # We will use the Shared RNG variable directly because Scan demands it, even
+    # though it would not be necessary for the icdf OgFromGraph
+    graph_inputs = [*rv.owner.inputs[1:], lower, upper]
+    graph_inputs_ = [inp.type() for inp in graph_inputs]
+    size_, dtype_, *rv_inputs_, lower_, upper_ = graph_inputs_
+    rv_ = srng.gen(rv.owner.op, *rv_inputs_, size=size_, dtype=dtype_)
+
+    # Try to use inverted cdf sampling
+    try:
+        # For left truncated discrete RVs, we need to include the whole lower bound.
+        # This may result in draws below the truncation range, if any uniform == 0
+        lower_value = lower_ - 1 if rv.owner.op.dtype.startswith("int") else lower_
+        cdf_lower_ = at.exp(logcdf(rv_, lower_value))
+        cdf_upper_ = at.exp(logcdf(rv_, upper_))
+        uniform_ = srng.uniform(
+            cdf_lower_,
+            cdf_upper_,
+            size=size_,
+        )
+        truncated_rv_ = icdf(rv_, uniform_)
+        truncated_rv = TruncatedRV(
+            base_rv_op=rv.owner.op,
+            inputs=graph_inputs_,
+            outputs=[uniform_.owner.outputs[0], truncated_rv_],
+            inline=True,
+        )(*graph_inputs)
+        updates = (
+            truncated_rv.owner.inputs[-1],
+            truncated_rv.owner.outputs[0],
+        )
+        return truncated_rv, updates
+    except NotImplementedError:
+        pass
+
+    # Fallback to rejection sampling
+    # TODO: Handle potential broadcast by lower / upper
+    def loop_fn(truncated_rv, reject_draws, lower, upper, rng, *rv_inputs):
+        next_rng, new_truncated_rv = rv.owner.op.make_node(rng, *rv_inputs).outputs
+        truncated_rv = at.set_subtensor(
+            truncated_rv[reject_draws],
+            new_truncated_rv[reject_draws],
+        )
+        reject_draws = at.or_((truncated_rv < lower), (truncated_rv > upper))
+
+        return (
+            (truncated_rv, reject_draws),
+            [(rng, next_rng)],
+            until(~at.any(reject_draws)),
+        )
+
+    # We want to use the same rng for the initial state and scan steps
+    # TODO: Will the default_updates interfere with the scan updates?
+    rng = rv_.owner.inputs[0]
+
+    (truncated_rv_, reject_draws_), updates = scan(
+        loop_fn,
+        outputs_info=[
+            at.zeros_like(rv_),
+            at.ones_like(rv_, dtype=bool),
+        ],
+        non_sequences=[lower_, upper_, rng, size_, dtype_, *rv_inputs_],
+        n_steps=max_n_steps,
+        strict=True,
+    )
```

Comment:
Use `srng.gen` inside of `loop_fn`, then you can remove all of the `rng` arguments between `loop_fn` and `scan` and the updates return value in `loop_fn`.

Reply:
I did not do that because it would return a graph with two rngs, one for the `zeros_like` and one for the body inner loop.

Reply:
> I did not do that because it would return a graph with two rngs, one for the `zeros_like` and one for the body inner loop.

Wherever `RandomStream.gen` is used a new RNG object will be created.  There's nothing inherently bad about that, though; however, the current approach has all the same inconsistencies I mentioned earlier.